File: packages/core-legacy/lib/target_resolver.ts
import type {
  KatmerConfig,
  KatmerHostResolved,
  KatmerHostInput
} from "./interfaces/config.interface"
import { toMerged } from "es-toolkit"
import type { KatmerProvider } from "./interfaces/provider.interface"
import objectHash from "stable-hash"
import { SSHProvider } from "./providers/ssh/ssh.provider"
import type { KatmerCore } from "./katmer"
import { LocalProvider } from "./providers/local.provider"
import { wildcardMatch } from "./utils/string.utils"
import { evalObjectVals, evalTemplate } from "./utils/renderer/renderer"

export class KatmerTargetResolver {
  #providerCache = new Map<string, KatmerProvider>()

  #allNames: Set<string>
  #groups: Map<string, Set<string>>
  #hosts: Map<string, KatmerHostResolved>
  constructor(
    private core: KatmerCore,
    ...targets: KatmerConfig["targets"][]
  ) {
    const normalized = KatmerTargetResolver.normalizeHosts(
      this.core.config.targets,
      ...targets
    )
    this.#hosts = normalized.hosts
    this.#groups = normalized.groups
    this.#allNames = normalized.allNames
  }

  get hosts() {
    return this.#hosts
  }

  async resolveProvider(opts: KatmerHostResolved): Promise<KatmerProvider> {
    const key = objectHash(opts)

    if (this.#providerCache.has(key)) {
      return this.#providerCache.get(key)!
    }
    let provider: KatmerProvider
    if (!opts.connection || opts.connection === "ssh") {
      provider = new SSHProvider(opts)
    } else if (opts.connection === "local") {
      provider = new LocalProvider(opts)
    } else {
      throw new Error(`Unknown connection type: ${opts["connection"]}`)
    }

    provider.logger = this.core.logger.child({
      provider: provider.constructor.name
    })

    if (opts.variables) {
      provider.variables = toMerged(provider.variables, opts.variables)
    }
    if (opts.environment) {
      const env = toMerged(provider.environment, opts.environment)

      provider.environment = await evalObjectVals(env, {
        //   TODO: env context
      })
    }

    this.#providerCache.set(key, provider)
    return provider
  }

  resolveTargets(pattern: string) {
    if (pattern === "all" || pattern === "*") {
      return [...this.#hosts.values()]
    }

    const parts = pattern.split(/[:,]/)
    const included = [] as string[]
    const excluded = [] as string[]
    const intersected = [] as string[]

    for (const part of parts) {
      if (part.startsWith("!")) {
        const plainName = part.slice(1)
        if (excluded.indexOf(plainName) === -1) {
          excluded.push(plainName)
        }
      } else if (part.startsWith("@")) {
        const plainName = part.slice(1)
        if (intersected.indexOf(plainName) === -1) {
          intersected.push(part.slice(1))
        }
      } else if (included.indexOf(part) === -1) {
        const token = part === "all" ? "*" : part // ← normalize 'all'
        if (!included.includes(token)) included.push(token)
      }
    }
    const wildcard = (name: string, pat: string) => wildcardMatch(name, pat)
    const matchesAny = (name: string, list: string[]) =>
      list.some((p) => wildcard(name, p))
    const isExcluded = (name: string) => excluded.some((p) => wildcard(name, p))

    // Stage 1: choose label candidates (hosts or groups), honoring exclusion
    const candidateLabels = new Set<string>()
    for (const name of this.#allNames) {
      if (isExcluded(name)) continue
      if (included.length === 0 || matchesAny(name, included)) {
        candidateLabels.add(name)
      }
    }

    // Stage 2: expand labels to hostnames, honoring exclusion & dedupe
    const expandedHostnames = new Set<string>()
    for (const label of candidateLabels) {
      if (this.#groups.has(label)) {
        for (const host of this.#groups.get(label)!) {
          if (!isExcluded(host)) expandedHostnames.add(host)
        }
      } else if (this.#hosts.has(label)) {
        if (!isExcluded(label)) expandedHostnames.add(label)
      }
    }

    // Stage 3: optional intersection (@foo) applied on final hostnames
    let finalHostnames = Array.from(expandedHostnames)
    if (intersected.length > 0) {
      finalHostnames = finalHostnames.filter((h) => matchesAny(h, intersected))
    }

    const resolved = finalHostnames
      .map((h) => this.#hosts.get(h)!)
      .filter(Boolean)

    if (resolved.length === 0) {
      throw new Error(`No targets found for pattern: ${pattern}`)
    }
    return resolved
  }

  static normalizeHosts(...inputs: KatmerConfig["targets"][]) {
    const reservedKeys = [
      "all",
      "children",
      "settings",
      "variables",
      "environment",
      "hosts"
    ]
    const allNames = new Set<string>()
    const hosts = new Map<string, KatmerHostResolved>()
    const groups = new Map<string, Set<string>>()

    const groupSettingsAccum = new Map<string, Record<string, any>>()
    const groupVariablesAccum = new Map<string, Record<string, any>>()
    const groupEnvAccum = new Map<string, Record<string, any>>() // NEW

    for (const input of inputs) {
      if (!input) continue

      const rootKeys = Object.keys(input)
      const isHostsDef =
        rootKeys.includes("hosts") ||
        rootKeys.includes("settings") ||
        rootKeys.includes("variables") ||
        rootKeys.includes("environment") // NEW

      if (isHostsDef) {
        // Ungrouped (root) settings
        const incomingSettings = (input as any).settings || {}
        const prevSettings = groupSettingsAccum.get("ungrouped") || {}
        const effSettings = toMerged(prevSettings, incomingSettings) as any
        groupSettingsAccum.set("ungrouped", effSettings)

        // Ungrouped variables
        const incomingVars = (input as any).variables || {}
        const prevVars = groupVariablesAccum.get("ungrouped") || {}
        const effVars = toMerged(prevVars, incomingVars) as any
        groupVariablesAccum.set("ungrouped", effVars)

        // Ungrouped environment (NEW)
        const incomingEnv = (input as any).environment || {}
        const prevEnv = groupEnvAccum.get("ungrouped") || {}
        const effEnv = toMerged(prevEnv, incomingEnv) as any
        groupEnvAccum.set("ungrouped", effEnv)

        processHostEntries(
          (input as any).hosts,
          "ungrouped",
          effSettings,
          effVars,
          effEnv
        )
      } else {
        for (const [groupName, def] of Object.entries(input)) {
          if (reservedKeys.includes(groupName)) {
            throw `cannot use '${groupName}' as group name: it is a reserved keyword`
          }

          // Group settings
          const incomingSettings = (def as any)?.settings || {}
          const prevSettings = groupSettingsAccum.get(groupName) || {}
          const effSettings = toMerged(prevSettings, incomingSettings) as any
          groupSettingsAccum.set(groupName, effSettings)

          // Group variables
          const incomingVars = (def as any)?.variables || {}
          const prevVars = groupVariablesAccum.get(groupName) || {}
          const effVars = toMerged(prevVars, incomingVars) as any
          groupVariablesAccum.set(groupName, effVars)

          // Group environment (NEW)
          const incomingEnv = (def as any)?.environment || {}
          const prevEnv = groupEnvAccum.get(groupName) || {}
          const effEnv = toMerged(prevEnv, incomingEnv) as any
          groupEnvAccum.set(groupName, effEnv)

          processHostEntries(
            def?.hosts,
            groupName,
            effSettings,
            effVars,
            effEnv
          )

          // Link children: inherit parent *settings/variables/environment* to child hosts
          const childNames = Object.keys(def?.children || {})
          for (const childGroupName of childNames) {
            if (!groups.has(childGroupName)) {
              throw `child group not found: '${childGroupName}' in group: '${groupName}'`
            }
            for (const child_hostname of groups.get(childGroupName)!) {
              const prevChild = (hosts.get(child_hostname) || {}) as any
              const childVarsPrev = prevChild.variables || {}
              const childEnvPrev = prevChild.environment || {}
              const parentVars = groupVariablesAccum.get(groupName) || {}
              const parentEnv = groupEnvAccum.get(groupName) || {}

              hosts.set(child_hostname, {
                ...(toMerged(prevChild, effSettings) as any),
                name: child_hostname,
                variables: toMerged(childVarsPrev, parentVars),
                environment: toMerged(childEnvPrev, parentEnv) // NEW
              })
            }
          }
        }
      }
    }

    function processHostEntries(
      hostConfig?: Record<string, KatmerHostInput>,
      groupName: string = "ungrouped",
      groupSettings: Record<string, any> = {},
      groupVariables: Record<string, any> = {},
      groupEnvironment: Record<string, any> = {} // NEW
    ) {
      if (!groups.has(groupName)) groups.set(groupName, new Set<string>())
      const groupHosts = groups.get(groupName)!
      allNames.add(groupName)

      for (const [hostname, host_settings] of Object.entries(
        hostConfig || {}
      )) {
        if (reservedKeys.includes(hostname)) {
          throw `cannot use '${hostname}' as hostname: it is a reserved keyword`
        }
        allNames.add(hostname)
        groupHosts.add(hostname)

        const prev = (hosts.get(hostname) || {}) as any
        const mergedConn = toMerged(
          prev,
          toMerged(groupSettings || {}, host_settings || {})
        ) as any

        const prevVars = prev.variables || {}
        const prevEnv = prev.environment || {}

        hosts.set(hostname, {
          ...mergedConn,
          name: hostname,
          variables: toMerged(prevVars, groupVariables || {}),
          environment: toMerged(prevEnv, groupEnvironment || {}) // NEW
        })
      }
    }

    return { groups, hosts, allNames }
  }

  async [Symbol.asyncDispose]() {
    try {
      for (const [_hash, provider] of this.#providerCache.entries()) {
        await provider.safeShutdown()
      }
    } catch {}
  }
}

File: packages/core/src/target_resolver.rs
use crate::config::{KatmerConfig, KatmerHostResolved, Targets, HostInput, Group, RootTargets};
use crate::utils::string::wildcard_match;
use std::collections::{HashMap, HashSet};
use serde_json::Value;

pub struct KatmerTargetResolver {
    all_names: HashSet<String>,
    groups: HashMap<String, HashSet<String>>,
    hosts: HashMap<String, KatmerHostResolved>,
}

impl KatmerTargetResolver {
    pub fn new(config: &KatmerConfig) -> Self {
        let (groups, hosts, all_names) = Self::normalize_hosts(&config.targets);
        Self {
            all_names,
            groups,
            hosts,
        }
    }

    pub fn resolve_targets(&self, pattern: &str) -> Vec<KatmerHostResolved> {
        if pattern == "all" || pattern == "*" {
            return self.hosts.values().cloned().collect();
        }

        let parts: Vec<&str> = pattern.split(|c| c == ':' || c == ',').collect();
        let mut included = Vec::new();
        let mut excluded = Vec::new();
        let mut intersected = Vec::new();

        for part in parts {
            let part = part.trim();
            if part.is_empty() { continue; }

            if part.starts_with('!') {
                excluded.push(&part[1..]);
            } else if part.starts_with('@') {
                intersected.push(&part[1..]);
            } else {
                let token = if part == "all" { "*" } else { part };
                included.push(token);
            }
        }

        let is_excluded = |name: &str| excluded.iter().any(|p| wildcard_match(name, p));
        let matches_any = |name: &str, list: &[&str]| list.iter().any(|p| wildcard_match(name, p));

        // choose label candidates (hosts or groups), honoring exclusion
        let mut candidate_labels = HashSet::new();
        for name in &self.all_names {
            if is_excluded(name) {
                continue;
            }
            if included.is_empty() || matches_any(name, &included) {
                candidate_labels.insert(name);
            }
        }

        // expand labels to hostnames, honoring exclusion & dedupe
        let mut expanded_hostnames = HashSet::new();
        for label in candidate_labels {
            if let Some(group_hosts) = self.groups.get(label) {
                for host in group_hosts {
                    if !is_excluded(host) {
                        expanded_hostnames.insert(host);
                    }
                }
            } else if self.hosts.contains_key(label) {
                if !is_excluded(label) {
                    expanded_hostnames.insert(label);
                }
            }
        }

        // optional intersection (@foo) applied on final hostnames
        let mut final_hostnames: Vec<&String> = expanded_hostnames.into_iter().collect();
        if !intersected.is_empty() {
            final_hostnames.retain(|h| matches_any(h, &intersected));
        }

        let mut result: Vec<KatmerHostResolved> = final_hostnames
            .iter()
            .filter_map(|h| self.hosts.get(*h))
            .cloned()
            .collect();

        if result.is_empty() && !pattern.is_empty() {
            // TODO: check mode / log warning or return error 
        }

        result
    }

    fn normalize_hosts(targets: &Targets) -> (HashMap<String, HashSet<String>>, HashMap<String, KatmerHostResolved>, HashSet<String>) {
        let mut all_names = HashSet::new();
        let mut hosts = HashMap::new();
        let mut groups: HashMap<String, HashSet<String>> = HashMap::new();

        match targets {
            Targets::Root(root) => {
                Self::process_root(root, &mut all_names, &mut hosts, &mut groups);
            }
            Targets::Grouped(grouped) => {
                let mut group_settings = HashMap::new();
                let mut group_variables = HashMap::new();
                let mut group_env = HashMap::new();

                // First pass: resolve basic groups and hosts
                for (group_name, group_def) in grouped {
                    Self::process_group(
                        group_name,
                        group_def,
                        &mut all_names,
                        &mut hosts,
                        &mut groups,
                        &mut group_settings,
                        &mut group_variables,
                        &mut group_env
                    );
                }

                //  handle children inheritance
                for (group_name, group_def) in grouped {
                    if let Some(children) = &group_def.children {
                        for child_group_name in children.keys() {
                            if let Some(child_hosts) = groups.get(child_group_name).cloned() {
                                let parent_settings = group_settings.get(group_name).cloned().unwrap_or_default();
                                let parent_vars = group_variables.get(group_name).cloned().unwrap_or_default();
                                let parent_env = group_env.get(group_name).cloned().unwrap_or_default();

                                for child_hostname in child_hosts {
                                    if let Some(host_resolved) = hosts.get_mut(&child_hostname) {
                                        // Merge parent settings into connection
                                        host_resolved.connection = Self::merge_host_input(&host_resolved.connection, &parent_settings);
                                        // Merge parent variables and env
                                        host_resolved.variables = Self::merge_maps(&host_resolved.variables, &parent_vars);
                                        host_resolved.environment = Self::merge_string_maps(&host_resolved.environment, &parent_env);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        (groups, hosts, all_names)
    }

    fn process_root(root: &RootTargets, all_names: &mut HashSet<String>, hosts: &mut HashMap<String, KatmerHostResolved>, groups: &mut HashMap<String, HashSet<String>>) {
        let group_name = "ungrouped";
        all_names.insert(group_name.to_string());
        let mut group_hosts = HashSet::new();

        let root_settings = root.settings.clone().unwrap_or_default();
        let root_vars = root.variables.clone().unwrap_or_default();
        let root_env = root.environment.clone().unwrap_or_default();

        for (hostname, input) in &root.hosts {
            all_names.insert(hostname.clone());
            group_hosts.insert(hostname.clone());

            let merged_input = Self::merge_host_input(input, &root_settings);
            
            let mut host_vars = root_vars.clone();
            host_vars.insert("inventory_hostname".to_string(), serde_json::Value::String(hostname.clone()));

            let resolved = KatmerHostResolved {
                name: hostname.clone(),
                connection: merged_input,
                variables: host_vars,
                environment: root_env.clone(),
            };
            hosts.insert(hostname.clone(), resolved);
        }
        groups.insert(group_name.to_string(), group_hosts);
    }

    fn process_group(
        group_name: &str,
        group_def: &Group,
        all_names: &mut HashSet<String>,
        hosts: &mut HashMap<String, KatmerHostResolved>,
        groups: &mut HashMap<String, HashSet<String>>,
        group_settings_accum: &mut HashMap<String, HashMap<String, Value>>,
        group_variables_accum: &mut HashMap<String, HashMap<String, Value>>,
        group_env_accum: &mut HashMap<String, HashMap<String, String>>,
    ) {
        all_names.insert(group_name.to_string());
        groups.entry(group_name.to_string()).or_insert_with(HashSet::new);

        let settings = group_def.settings.clone().unwrap_or_default();
        let variables = group_def.variables.clone().unwrap_or_default();
        let environment = group_def.environment.clone().unwrap_or_default();

        group_settings_accum.insert(group_name.to_string(), settings.clone());
        group_variables_accum.insert(group_name.to_string(), variables.clone());
        group_env_accum.insert(group_name.to_string(), environment.clone());

        if let Some(host_map) = &group_def.hosts {
            let group_hosts = groups.get_mut(group_name).unwrap();
            for (hostname, input) in host_map {
                all_names.insert(hostname.clone());
                group_hosts.insert(hostname.clone());

                let merged_input = Self::merge_host_input(input, &settings);

                let mut host_vars = variables.clone();
                host_vars.insert("inventory_hostname".to_string(), serde_json::Value::String(hostname.clone()));

                let resolved = KatmerHostResolved {
                    name: hostname.clone(),
                    connection: merged_input,
                    variables: host_vars,
                    environment: environment.clone(),
                };
                hosts.insert(hostname.clone(), resolved);
            }
        }
    }

    fn merge_host_input(base: &HostInput, settings: &HashMap<String, Value>) -> HostInput {
        if settings.is_empty() {
            return base.clone();
        }

        let mut base_val = serde_json::to_value(base).unwrap();
        if let Value::Object(ref mut map) = base_val {
            for (k, v) in settings {
                map.insert(k.clone(), v.clone());
            }
        }
        
        serde_json::from_value(base_val).unwrap_or_else(|_| base.clone())
    }

    fn merge_maps(base: &HashMap<String, Value>, incoming: &HashMap<String, Value>) -> HashMap<String, Value> {
        let mut result = base.clone();
        for (k, v) in incoming {
            result.insert(k.clone(), v.clone()); // Shadow base if same key
        }
        result
    }

    fn merge_string_maps(base: &HashMap<String, String>, incoming: &HashMap<String, String>) -> HashMap<String, String> {
        let mut result = base.clone();
        for (k, v) in incoming {
            result.insert(k.clone(), v.clone());
        }
        result
    }
}

